src/libexpr/include/nix/expr/nixexpr.hh

```diff
@@ -365,74 +365,273 @@ struct ExprOpHasAttr : Expr
-    };
 
-    typedef std::map<Symbol, AttrDef> AttrDefs;
-    AttrDefs attrs;
-    std::unique_ptr<std::vector<Expr *>> inheritFromExprs;
+        std::ranges::copy(builder->inheritFromExprs, inheritFromExprsStart);
 
-    struct DynamicAttrDef
-    {
-        Expr *nameExpr, *valueExpr;
-        PosIdx pos;
-        DynamicAttrDef(Expr * nameExpr, Expr * valueExpr, const PosIdx & pos)
-            : nameExpr(nameExpr)
-            , valueExpr(valueExpr)
-            , pos(pos) {};
-    };
+        for (auto const & [n, dynAttr] : enumerate(builder->dynamicAttrs))
+            dynamicAttrsStart[n] = DynamicAttrDef(alloc, dynAttr);
 
-    typedef std::vector<DynamicAttrDef> DynamicAttrDefs;
-    DynamicAttrDefs dynamicAttrs;
+        delete builder;
+    }
+
+    /**
+     * We can skip the circus if the attrs is empty
+     */
     ExprAttrs(const PosIdx & pos)
         : recursive(false)
-        , pos(pos) {};
+        , pos(pos)
+        , nAttrs(0)
+        , nInheritFromExprs(0)
+        , nDynamicAttrs(0)
+        , attrNamesStart(nullptr)
+        , attrValuesStart(nullptr)
+        , inheritFromExprsStart(nullptr)
+        , dynamicAttrsStart(nullptr)
+    {
+    }
+
     ExprAttrs()
-        : recursive(false) {};
+        : ExprAttrs(noPos)
+    {
+    }
+
 
     PosIdx getPos() const override
     {
         return pos;
     }
 
+    std::span<Symbol> getAttrNames() const {
+        return {attrNamesStart, nAttrs};
+    }
+
+    std::span<AttrDef> getAttrValues() const {
+        return {attrValuesStart, nAttrs};
+    }
+
+    std::span<Expr *> getInheritFromExprs() const
+    {
+        return {inheritFromExprsStart, nInheritFromExprs};
+    }
+
+    std::span<DynamicAttrDef> getDynamicAttrs() const
+    {
+        return {dynamicAttrsStart, nDynamicAttrs};
+    }
+
+    /**
+     * Association map for looking up a value by name, or iterating over all
+     * (name, value) pairs. This is meant to mimic what std::map can do, but
+     * without the extra overhead necessary to be mutable. The layout does not
+     * need to be optimized as it is created temporarily on-demand
+     */
+    struct AttrDefs {
+        uint32_t nAttrs;
+        Symbol * attrNamesStart;
+        AttrDef * attrValuesStart;
+
+        AttrDef * find(Symbol s)
+        {
+            auto attrNamesEnd = attrNamesStart + nAttrs;
+            auto found = std::lower_bound(attrNamesStart, attrNamesEnd, s);
+
+            if (found == attrNamesEnd || *found != s)
+                return nullptr;
+
+            auto idx = found - attrNamesStart;
+            return &attrValuesStart[idx];
+
+        }
+
+        inline size_t displ(AttrDef * attr)
+        {
+            assert(attrValuesStart <= attr && attr < attrValuesStart + nAttrs);
+
+            return attr - attrValuesStart;
+        }
+
+        size_t size()
+        {
+            return nAttrs;
+        }
+
+        // XXX: help I don't know how to C++ I'm making this up
+        class iterator
```

Comment:
Someone please help with proper iterator design I was so confused trying to figure this out